parser: Allow parsing of qualified type path as nested generic argument

CohenArthur opened the pull request:
Let's take the example of lexing `Option<<T as Iterator>::Item>` and look at the first few tokens. Originally, `Option<<T` was lexed as 3 tokens:

* IDENTIFIER(Option)
* LEFT_SHIFT
* IDENTIFIER(T)

The parser did not allow a list of generic arguments to start with a left shift, and rejected the above type. We are now splitting the left shift into two left angles, as this allows complex generic arguments and overall makes sense parsing wise. Thus, the above list becomes:

* IDENTIFIER(Option)
* LEFT_ANGLE
* LEFT_ANGLE
* IDENTIFIER(T)

and `<T as Iterator>` is properly parsed as a qualified path.

Fixes #1815
Fixed #1809

Addresses #1524

gcc/rust/ChangeLog:

	* parse/rust-parse-impl.h (Parser::parse_path_generic_args): Split leading `LEFT_SHIFT` token into two `LEFT_ANGLE` tokens when parsing generic arguments. (Parser::parse_type_path_segment): Allow `LEFT_ANGLE` as starting token for parsing generic arguments.

gcc/testsuite/ChangeLog:

	* rust/compile/parse_associated_type_as_generic_arg.rs: New test.
	* rust/compile/parse_associated_type_as_generic_arg2.rs: New test.
	* rust/compile/path_as_generic_arg.rs: New test.
	* rust/compile/nested_generic.rs: New test.

@sun-jacobi I accidentally fixed your issue while investigating something different. Sorry :(

P-E-P commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -6375,6 +6375,9 @@ template <typename ManagedTokenSource>
 AST::GenericArgs
 Parser<ManagedTokenSource>::parse_path_generic_args ()
 {
+  if (lexer.peek_token ()->get_id () == LEFT_SHIFT)
```
I think we cannot simply peek a token here and expect it to be either a left shift or a left angle, this will not scale with deeply nested generics.

P-E-P commented:
The parser fail with the following snippet @CohenArthur :

<details>
  <summary>Failing code(Click me)</summary>

```rust
trait Bar {
    type B;

    fn bar();
}

trait Foo {
    type A;

    fn foo();
}

trait Toto {
    type C;

    fn toto();
}

trait Tata {
    type D;
    fn tata();
}

impl Toto for u32 {
    type C = f32;

    fn toto() {}
}

impl Tata for f32 {
    type D = u32;

    fn tata() {}
}

struct S;

impl Bar for i32 {
    type B = u32;

    fn bar() {}
}

impl Foo for S {
    type A = i32;

    fn foo() {}
}

enum Maybe<T> {
    Something(T),
    Nothing,
}

fn foo() -> Maybe<<<<<S as Foo>::A as Bar>::B as Toto>::C as Tata>::D> {
    Maybe::Something(15)
}
```
</details>

This is simply one of your test extended with even more nesting. It does work on [rustc](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8f087eed7679c2ab48b159b4c3b7e9d8).

CohenArthur commented:
> The parser fail with the following snippet @CohenArthur :
> Failing code(Click me)
> 
> This is simply one of your test extended with even more nesting. It does work on [rustc](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8f087eed7679c2ab48b159b4c3b7e9d8).

Very good catch :D This is now fixed and I added your snippet as a test case. Thank you :)

philberty commented:
I cant wait to look fixing the other ICE we have relating to this :D

philberty reviewed (approved):
LGTM its a tricky one this issue

CohenArthur commented:
> I cant wait to look fixing the other ICE we have relating to this :D

Yeah, this is why I'm annotating the testcases with `-fsyntax-only`. Even if we can't yet understand those types at least we can parse them and you can get started on fixing the big issue :D
